godot-core/src/builtin/projection.rs

```diff
@@ -460,3 +509,474 @@ pub enum ProjectionEye {
     Left = 1,
     Right = 2,
 }
+
+#[cfg(test)]
+mod test {
+    #![allow(clippy::type_complexity, clippy::excessive_precision)]
+
+    use crate::assert_eq_approx;
+
+    use super::*;
+
+    const EPSILON: real = 1e-6;
+
+    fn real_is_approx(a: real, b: real) -> bool {
+        (a - b).abs() <= EPSILON
+    }
+
+    fn matrix_eq_approx(a: Projection, b: RMat4) -> bool {
+        a.to_glam().abs_diff_eq(b, EPSILON)
+    }
+
+    /// Test that diagonals matrices has certain property.
+    #[test]
+    fn test_diagonals() {
+        const DIAGONALS: [[real; 4]; 10] = [
+            [1.0, 1.0, 1.0, 1.0],
+            [2.0, 1.0, 2.0, 1.0],
+            [3.0, 2.0, 1.0, 1.0],
+            [-1.0, -1.0, 1.0, 1.0],
+            [0.0, 0.0, 0.0, 0.0],
+            [-2.0, -3.0, -4.0, -5.0],
+            [0.0, 5.0, -10.0, 50.0],
+            [-1.0, 0.0, 1.0, 100.0],
+            [-15.0, -22.0, 0.0, 11.0],
+            [-1.0, 3.0, 1.0, 0.0],
+        ];
+
+        for [x, y, z, w] in DIAGONALS {
+            let proj = Projection::from_diagonal(x, y, z, w);
+            assert_eq_approx!(
+                proj,
+                RMat4::from_cols_array(&[
+                    x, 0.0, 0.0, 0.0, 0.0, y, 0.0, 0.0, 0.0, 0.0, z, 0.0, 0.0, 0.0, 0.0, w,
+                ]),
+                matrix_eq_approx,
+            );
+
+            let det = x * y * z * w;
+            assert_eq_approx!(proj.determinant(), det, real_is_approx);
+            if det.abs() > 1e-6 {
+                assert_eq_approx!(
+                    proj.inverse(),
+                    RMat4::from_cols_array(&[
+                        1.0 / x,
+                        0.0,
+                        0.0,
+                        0.0,
+                        0.0,
+                        1.0 / y,
+                        0.0,
+                        0.0,
+                        0.0,
+                        0.0,
+                        1.0 / z,
+                        0.0,
+                        0.0,
+                        0.0,
+                        0.0,
+                        1.0 / w,
+                    ]),
```

Comment:
Maybe that's a case for `#[rustfmt::skip]` and 4-by-4 matrix notation for readability. I think this has been done elsewhere already.

Maybe use a separate variable:
```rs
 if det.abs() > 1e-6 {
    #[rustfmt::skip]
    let expected = RMat4::from_cols_array(&[
        // 4 rows, 4 columns
    ]);
    assert_eq_approx!(proj.inverse(), expected);
```